Moved login/logout logic into the useUser composable and updated references

iamshobhraj opened the pull request:
## Summary
<!--
 * description of the change
 * manual verification steps performed
 * screenshots if the PR affects the UI
-->

Login/logout and session logic is moved into `useUser` composable from `action.js`. Refactored vuex state into module level state. References to `CORE_SET_SESSION` is updated to use the setSession method from useUser composable.  

## References
<!--
 * references to related issues and PRs
 * links to mockups or specs for new features
 * links to the diffs for any dependency updates, e.g. in iceqube or the perseus plugin
-->
fixes #12204 

## Reviewer guidance
<!--
 * how can a reviewer test these changes?
 * are there any risky areas that deserve extra testing
-->

after updating references to `CORE_SET_SESSION` in `heartbeat.spec.js` file, some tests still failed. I am unable to figure out why these tests are failing. I updated the mock file and implementated it as written in documentation.

nucleogenesis commented on packages/kolibri/__tests__/heartbeat.spec.js:
```diff
@@ -1,27 +1,25 @@
   stubWindowLocation(beforeAll, afterAll);
   // replace the real XHR object with the mock XHR object before each test
-  beforeEach(() => mock.setup());
+  beforeEach(() => useUser.mockImplementation(() => useUserMock()));
```
Is there a particular reason for `mock.setup()` not to be called anymore? I'm wondering if this is related to the failing tests

iamshobhraj replied:
Sorry for the oversight. I’ve fixed this now.

nucleogenesis reviewed (changes requested):
In the description you mentioned that the logic was moved _from_ an actions.js file but there are no changes to any Vuex files included.

I also see that `heartbeat.js` may need to be updated to use the `useUser` module as it is still calling `dispatch('setSession', ...)` [here](https://github.com/iamshobhraj/kolibri/blob/b5647baeb48179b37e234489e68bbe33bfde23f8/packages/kolibri/heartbeat.js#L199).

I had a comment in the heartbeat.spec.js file re: the tests but I think that updating `heartbeat.js` might be the key

iamshobhraj commented:
> In the description you mentioned that the logic was moved _from_ an actions.js file but there are no changes to any Vuex files included.
> 
@nucleogenesis Thanks for pointing that out. I had initially planned on removing the Vuex logic after successfully migrating the login and session logic to the new Composition API. However, I got stuck with the heartbeat.spec.js tests. I’ve now removed the login/logout and session logic from the Vuex files.

> I also see that `heartbeat.js` may need to be updated to use the `useUser` module as it is still calling `dispatch('setSession', ...)` [here](https://github.com/iamshobhraj/kolibri/blob/b5647baeb48179b37e234489e68bbe33bfde23f8/packages/kolibri/heartbeat.js#L199).
> 
> I had a comment in the heartbeat.spec.js file re: the tests but I think that updating `heartbeat.js` might be the key
>

Regarding the heartbeat.js, I refactored the store.dispatch to use the setSession method, but the failing tests persist. I suspect the issue might be related to xhr-mock, but I’m unable to pinpoint the exact cause.

Please let me know if you have any suggestions.

nucleogenesis commented on packages/kolibri/composables/__mocks__/useUser.js:
```diff
@@ -77,7 +78,29 @@ export function useUserMock(overrides = {}) {
+    session, // Make session mutable for test scenarios
+
+    // Actions
+    setSession: jest.fn(({ session: newSession, clientNow }) => {
```
One of the failing tests re: not calling `serverTime` w/ `clientNow` may be fixed by calling it here.

nucleogenesis reviewed (commented):
I added one comment to the mock function for `setSession` and after some thinking and looking at the code locally I'm not sure precisely what the fix is, but I'm feeling it is very likely to do with that function altogether and it's differences to the real `setSession`.

Another thing I saw was that the `kolibriLogin` action is being dispatched a few places - such as the [`store.spec.js` file](https://github.com/iamshobhraj/kolibri/blob/ec7851aa405df9e32f819ad44cd488128408854b/packages/kolibri/__tests__/store.spec.js#L61-L62) 

Thank you for your effort on this. I hope anything I've said can be helpful. I'll keep thinking about this and come back to it again tomorrow if I can

rtibbles commented:
When I did a rebase locally on develop here, it threw up some weird conflicts, but resolving them during the rebase did reduce it down to only about 10 commits, so I think this can be cleaned up! Let me know if it would be helpful for me to force push updates for this.

iamshobhraj commented:
> When I did a rebase locally on develop here, it threw up some weird conflicts, but resolving them during the rebase did reduce it down to only about 10 commits, so I think this can be cleaned up! Let me know if it would be helpful for me to force push updates for this.

Thank you for your help! If you think force pushing the updates would clean up the commit history and make it easier to manage, I would appreciate it. Please go ahead and do so if it's not too much trouble.

rtibbles commented:
Sure thing, @iamshobhraj - I will make the update for you!

rtibbles commented on packages/kolibri/composables/useUser.js:
```diff
@@ -52,15 +145,11 @@ export default function useUser() {
     getUserKind,
     userHasPermissions,
     session,
-    //state
```
Did you cross check that these are not used anywhere before deleting them?

rtibbles commented on packages/kolibri/composables/useUser.js:
```diff
@@ -1,39 +1,132 @@
+    };
+  }
+
+  async function kolibrisetUnspecifiedPassword({ username, password, facility }) {
```
This doesn't need to be in here, we can leave it in the actions.js instead.

rtibbles commented on packages/kolibri/heartbeat.js:
```diff
@@ -196,17 +196,8 @@ export class HeartBeat {
         }
-        store.dispatch('setSession', {
-          session,
-          // Calculate an approximation of the client 'now' that was simultaneous to the server
```
This explanatory comment should not be deleted.

rtibbles reviewed (changes requested):
There are still a few failing tests after the rebase, but the commit history seems a bit more coherent now.

iamshobhraj commented:
@nucleogenesis  i've hit a roadblock with these failing tests:

1. heartbeat › _checksession method › should sign out if an auto logout is detected
1. heartbeat › _checksession method › should redirect if a change in user is detected
   
   I am unable to figure out or debug the cause of these failing tests.
   
  Could you please help me understand:
   1. Why these specific tests are failing?
   2. What am I missing in my implementation?

rtibbles commented:
For most of the tests it seems that they are still trying to use `store.dispatch` to trigger the login action - as we have removed that action, this will have no effect! So moving those tests into tests for the useUser composable instead seems like the first step here.

iamshobhraj committed:
useuser2.0

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

nucleogenesis commented:
@iamshobhraj - will you be able to come back to work out the failing tests & linting? I looked at your last commit and it's quite large - did you do a substantial rewrite since the last review?

iamshobhraj commented:
@nucleogenesis I was trying to go back and do everything from scratch. Last time i was stuck on the failing test for `heartbeat.spec.js`, but this time ignored it initially and tried to update references in other files hoping that would fix the failing test. But the failing test in `heartbeat.spec.js` file is still persistent. From what i understand, the issue is in `_checksession` method particularly in disconnect-reconnect workflow. When Kolibri is disconnected and then reconnected to the server, the expected snackbar message is “Successfully reconnected,” but the test is throwing “Disconnected from the server.”. Same goes for when disconnected and trying to reconnect.

Also, I am sorry i havent been able to work on this pr over the past few weeks, I was occupied in some other stuff, particularly preparing my gsoc proposal, and i had hoped to return this some time later. So far i have moved the login/logout logic into useUser Composable and updated some references. But while updating the CORE_SET_SESSION referencee in `heartbeat.spec.js`, i think ive hit a roadblock.

iamshobhraj commented:
@nucleogenesis @rtibbles I'm closing this PR for now — I'll revisit it in the future after resolving the issue. In the meantime, feel free to reassign the issue if someone else would like to pursue it.
